src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -206,9 +206,20 @@ public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOExc
         }
     }
 
+    class ApplySettings implements NodeSettingsService.Listener {
+        @Override
+        public void onRefreshSettings(Settings settings) {
+            // Because the GroovyScriptEngineService knows nothing about the
+            // cache, we need to clear it here if the setting changes
+            if (settings.get(GroovyScriptEngineService.GROOVY_SCRIPT_BLACKLIST_PATCH) != null) {
+                ScriptService.this.clearCache();
```

Comment:
we also need to make sure we clear the cache *after* the settings have been updated on the engine itself. Maybe the engine won't registerer as a listener for settings on the service, and this class will be responsible to go over and call the refresh settings, and then call clear cache if needed?